SpaceTinder aka Matchmaking

comma opened the pull request:
This https://baystation12.net/forums/threads/free-poll-here-automating-cliques.3266/
Decided to put it up here so people who are not retarded / feverish can look at it too and point out where I derped up.
-Adds preferences for what random connections you wanna be eligible for
-Adds actual matchmaking between to form those connections
-Adds window to see the results and adjust them in case RNG did bad
-Propagation of info: other people knowing of your relations, personnel record data on people's relations.

Kelenius commented on code/modules/client/preference_setup/matchmaking/matchmaking.dm:
```diff
@@ -0,0 +1,151 @@
+			var/list/relations = matchmaker.get_relationships(src)
+			for(var/datum/relation/R in relations)
+				R.finalized = 1
+			usr << browse(null, "window=relations")
```
There actually is a macro for this, ``#define show_browser(target, browser_content, browser_name) target << browse(browser_content, browser_name)``

comma replied:
oh, didn't find ir in my cursory search, saw all the places using that still. Might change

Techhead0 commented on code/modules/client/preference_setup/matchmaking/relations_types.dm:
```diff
@@ -0,0 +1,41 @@
+	var/list/good = list()
+	for(var/datum/relation/R in rest)
+		var/datum/job/their = job_master.GetJob(R.holder.job)
+		var/datum/job/our = job_master.GetJob(holder.job)
```
Might be good to put this line outside the for loop, seeing as it will be the same every loop.

Techhead0 commented on code/modules/client/preference_setup/matchmaking/matchmaking.dm:
```diff
@@ -0,0 +1,151 @@
+/datum/relation/proc/get_candidates()
+	.= list()
+	for(var/datum/relation/R in matchmaker.relations)
+		if(R == src && R.holder == holder)
```
If `R == src`, then isn't `R.holder == holder` implicit?

comma replied:
Not really, matchmaker.relations holds ALL relations. So for example if dude has both Crossed and Was Crossed enabled, they could get paired up, and he'd cross himself.

Techhead0 replied:
Then wouldn't you want `if(R == src || R.holder == holder)`?

comma replied:
derp yeah

Techhead0 commented on code/modules/client/preference_setup/matchmaking/relations.dm:
```diff
@@ -0,0 +1,52 @@
+		if(R in pref.relations)
+			pref.relations -= R
+		else
+			pref.relations += R
```
You could just do `pref.relations ^= href_list["relation"]` if you wanted.

comma replied:
Witchery

comma committed:
Matchmaking

Adds randomly generated relationships between people.
You can opt in at char setup, select what kinds of relationsships your duder can have, and what info other party would have about them.
After you get matched to people, you can edit them once, either editing info tidbit or terminating relationship completely if RNG did bad.
Other people (mostly coworkers in same department) would know of what relationships you have.
They are also listed on your record in emlpoyment record consoles.

Techhead0 reviewed (approved):
Read through the PR, looks good, unless weird edge cases pop up.

Asanadas commented on code/modules/client/preference_setup/matchmaking/relations_types.dm:
```diff
@@ -0,0 +1,63 @@
+
+/datum/relation/rival
+	name = "Rival"
+	desc = "You and them are engaged in a constant struggle to show who's number one."
```
I believe it would be "You and they" here
>They and them are always used in place of plural nouns or noun groups in the third person. However the fundamental difference between the two in grammatical terms, is that they is a subject pronoun, and them is an object pronoun. They is used to refer to the subject of a clause.

but it's still onerous. How about "You both"?

Asanadas replied:
"You are engaged in a constant struggle with them to show who's number one"

"You are engaged with them in a constant struggle to show who's number one"

Is there like an [other.holder] gender pronoun get that could be plugged in here?

comma replied:
Not at this point, no, it's for preference setup window blurb mostly, so we don't know anything.

comma replied:
Also gonna steal idea from lower comment and just go with You are engaged etc

Asanadas commented on code/modules/client/preference_setup/matchmaking/relations_types.dm:
```diff
@@ -0,0 +1,63 @@
+
+/datum/relation/ex
+	name = "Ex"
+	desc = "You and them used to be romantically involved, but not anymore."
```
Same thing here, "You both".

Techhead0 replied:
`"You used to be romantically involved, but not anymore."` would be fine as well.

Asanadas reviewed (commented):
#justgrammarthings

babydoll commented on code/modules/client/preference_setup/matchmaking/matchmaking.dm:
```diff
@@ -0,0 +1,208 @@
+			R1.fields["connections"] |= get_desc_string()
+			R2.fields["connections"] |= get_desc_string()
+		else
+			R1.fields["connections"] |= "[holder] and [other.holder] know each other, exact nature of their relationship unclear."
```
X and Y know each other, **but the** exact nature of their relationship is unclear

babydoll commented on code/modules/client/preference_setup/matchmaking/matchmaking.dm:
```diff
@@ -0,0 +1,208 @@
+	if(href_list["info_relation"])
+		var/datum/relation/R = locate(href_list["info_relation"])
+		if(R)
+			var/info = sanitize(input("What would you like other party for this connection to know about your character?","Character info",R.info) as message|null)
```
What would you like **the** other party for this connection to know about your character?

babydoll commented on code/modules/client/preference_setup/matchmaking/matchmaking.dm:
```diff
@@ -0,0 +1,208 @@
+				return 1
+	if(href_list["relations_close"])
+		var/ok = "Close anyway"
+		ok = alert("HEY! You have some non-finalized relationships. You can terminate them if they do not fit your character, or edit info tidbit that other party is given. THIS IS YOUR ONLY CHANCE to do so - after you close the window, they won't be editable.","Finalize relationships","Return to edit", "Close anyway")
```
HEY! You have some non-finalized relationships. You can terminate them if they do not fit your character, or edit **the** info tidbit that **the** other party is given. THIS IS YOUR ONLY CHANCE to do so - after you close the window, they won't be editable."

babydoll commented on code/modules/client/preference_setup/matchmaking/relations.dm:
```diff
@@ -0,0 +1,49 @@
+		return TOPIC_REFRESH
+	if(href_list["relation_info"])
+		var/R = href_list["relation_info"]
+		var/info = sanitize(input("Character info", "What would you like other party for this connection to know about your character?",pref.relations_info[R]) as message|null)
```
What would you like **the** other party for this connection to know about your character?

babydoll commented on code/modules/client/preference_setup/matchmaking/relations_types.dm:
```diff
@@ -0,0 +1,63 @@
+/datum/relation/friend
+	name = "Friend"
+	desc = "You had known the fellow for a while now, and you get along pretty well."
```
You **have** known

babydoll commented on code/modules/client/preference_setup/matchmaking/relations_types.dm:
```diff
@@ -0,0 +1,63 @@
+
+/datum/relation/enemy
+	name = "Enemy"
+	desc = "You had known the fellow for a while now, and you really can't stand each other."
```
You **have** known

babydoll reviewed (commented):
Some more things to be changed if you want. I'm not sure if all of these are player facing.

PsiOmegaDelta commented:
There is really only one thing I note, there is no handling of mobs being gibbed (or otherwise deleted) causing the holder vars to become nulled.
The cleanest way to handle that might be to just change the holder var to /datum/mob_lite and grab the lite mob, assuming it has the values you need:
```holder = mob_repository.get_lite_mob(holder_mob)```

PsiOmegaDelta commented on code/modules/client/preferences.dm:
```diff
@@ -395,6 +399,14 @@ datum/preferences
+		for(var/T in relations)
+			var/TT = matchmaker.relation_types[T]
+			var/datum/relation/R = new TT
+			R.holder = character
```
Here is where the get-lite-call would be.
R.holder = mob_repository.get_lite_mob(character)

PsiOmegaDelta commented on code/modules/client/preference_setup/matchmaking/matchmaking.dm:
```diff
@@ -0,0 +1,208 @@
+	var/desc = "You just know the guy."
+	var/list/can_connect_to	//What relations (names) can matchmaking join us with? Defaults to own name.
+	var/list/incompatible 	//If we have relation like this with the mob, we can't join
+	var/mob/living/holder
```
And here is where holder becomes:
var/datum/mob_lite/holder

comma commented:
Turns out I can't use lite mob since I also need job and mind vars, that are not held by it.

comma commented:
Alright mind transition done.

Techhead0 reviewed (approved):
I read through it again after the mind changes, code looks good to me.

PsiOmegaDelta reviewed (approved):
I suppose we'll see if minds get deleted at some point and cause null runtimes anyway.

comma committed:
Merge remote-tracking branch 'refs/remotes/upstream/dev'

Conflicts:
	.travis.yml
